### PR TITLE
Add ROS2 version of Stopwatch

### PR DESCRIPTION
**Pull request by andermi**
# 🎉 New feature

## Summary
Add a ROS2 version of `math::Stopwatch` that can make use of `rclcpp::Clock` and sim time via `/clock`

## Test it
Check `test/test_stopwatch.cpp` for examples

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: ros_ign_gazebo/CMakeLists.txt [hunk 1]
```diff
@@ -74,12 +86,72 @@ install(FILES
...
+
+  # GTest
+  find_package(ament_cmake_gtest REQUIRED)
+  ament_find_gtest()
+
+  # launch_testing
+  find_package(launch_testing_ament_cmake REQUIRED)
+
+  # Helper function to generate gtest
+  function(gz_add_gtest TEST_NAME)
```
Comment (ahcorde) [inline ros_ign_gazebo/CMakeLists.txt]: I don't think it's a good idea to add this logic here, can you simplify this?

File: ros_ign_gazebo/include/ros_ign_gazebo/Stopwatch.hpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+  /// \brief Constructor.
+
+public:
+  Stopwatch();
+
+  /// \brief Copy constructor
+  /// \param[in] _watch The stop watch to copy.
+
+public:
+  Stopwatch(const Stopwatch & _watch);
```
Comment (ahcorde) [inline ros_ign_gazebo/include/ros_ign_gazebo/Stopwatch.hpp:60]: Please use this style
```suggestion
  /// \brief Copy constructor
  /// \param[in] _watch The stop watch to copy.
  public: Stopwatch(const Stopwatch & _watch);
```

**Review by quarkytale — APPROVED** [review]
lgtm

File: ros_ign_gazebo/CMakeLists.txt [hunk 3]
```diff
@@ -74,12 +86,72 @@ install(FILES
...
   ament_lint_auto_find_test_dependencies()
+
+  # GTest
+  find_package(ament_cmake_gtest REQUIRED)
+  ament_find_gtest()
+
+  # launch_testing
+  find_package(launch_testing_ament_cmake REQUIRED)
+
+  # Helper function to generate gtest
```
Comment (ahcorde) [inline ros_ign_gazebo/CMakeLists.txt]: simplify this logic

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: ros_ign_gazebo/include/ros_ign_gazebo/Stopwatch.hpp (hunk 2 above)
Comment (quarkytale) [reply ros_ign_gazebo/include/ros_ign_gazebo/Stopwatch.hpp:60]: Uncrustify is failing with that style though

File: ros_ign_gazebo/CMakeLists.txt [hunk 4]
```diff
@@ -57,6 +57,18 @@ target_link_libraries(create
...
+  rclcpp
+)
+target_include_directories(Stopwatch PUBLIC
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+  $<INSTALL_INTERFACE:include>)
+
+install(DIRECTORY include/${PROJECT_NAME}/
+  DESTINATION include/${PROJECT_NAME}
+)
+
```
Comment (ahcorde) [inline ros_ign_gazebo/CMakeLists.txt]:
```suggestion
add_library(Stopwatch src/Stopwatch.cpp)
ament_target_dependencies(Stopwatch
  rclcpp
)

target_include_directories(Stopwatch PUBLIC
  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
  $<INSTALL_INTERFACE:include>
)

install(DIRECTORY include/${PROJECT_NAME}/
  DESTINATION include/${PROJECT_NAME}
)

```

File: ros_ign_gazebo/CMakeLists.txt [hunk 5]
```diff
@@ -74,12 +86,35 @@ install(FILES
...
+    $<INSTALL_INTERFACE:include>)
+  target_link_libraries(test_stopwatch
+    Stopwatch
+  )
+  install(
+    TARGETS test_stopwatch
+    DESTINATION lib/${PROJECT_NAME}
+  )
+  ament_add_gtest_test(test_stopwatch)
+
```
Comment (ahcorde) [inline ros_ign_gazebo/CMakeLists.txt]:
```suggestion

  # GTest
  find_package(ament_cmake_gtest REQUIRED)
  ament_find_gtest()

  ament_add_gtest_executable(test_stopwatch
    test/test_stopwatch.cpp
  )
  
  ament_target_dependencies(test_stopwatch rclcpp)

  target_include_directories(test_stopwatch PUBLIC
      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
      $<INSTALL_INTERFACE:include>
   )

  target_link_libraries(test_stopwatch
    Stopwatch
  )

  install(
    TARGETS test_stopwatch
    DESTINATION lib/${PROJECT_NAME}
  )
  ament_add_gtest_test(test_stopwatch)

```

File: ros_ign_gazebo/include/ros_ign_gazebo/Stopwatch.hpp [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+  /// \brief Destructor.
+
+public:
+  virtual ~Stopwatch();
+
+public:
+  /// \brief Take a clock instance (e.g. get_clock() from rclcpp::Node).
+  /// Can also follow sim time on /clock when node's use_sim_time param is set
+  void SetClock(rclcpp::Clock::SharedPtr _clock);
```
Comment (ahcorde) [inline ros_ign_gazebo/include/ros_ign_gazebo/Stopwatch.hpp:77]:
```suggestion
    /// param[in] _clock
    void SetClock(rclcpp::Clock::SharedPtr _clock);
```

File: ros_ign_gazebo/src/Stopwatch.cpp [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+{
+  if (!this->dataPtr->clock) {
+    return false;
+  }
+
+  if (this->dataPtr->running) {
+    this->dataPtr->running = false;
+    this->dataPtr->stopTime = this->dataPtr->clock->now();
+    this->dataPtr->runDuration = this->dataPtr->runDuration + \
+      this->dataPtr->stopTime - this->dataPtr->startTime;
```
Comment (ahcorde) [inline ros_ign_gazebo/src/Stopwatch.cpp]:
```suggestion
    this->dataPtr->runDuration = this->dataPtr->runDuration +
      this->dataPtr->stopTime - this->dataPtr->startTime;
```

File: ros_ign_gazebo/test/test_stopwatch.cpp [hunk 8]
```diff
@@ -0,0 +1,398 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
```
Comment (ahcorde) [inline ros_ign_gazebo/test/test_stopwatch.cpp]:
```suggestion
// Copyright 2022 Open Source Robotics Foundation, Inc.
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Review by ahcorde — CHANGES_REQUESTED** [review]
Linters are failing 

```
4: /github/home/ws/src/ros_gz/ros_ign_gazebo/src/Stopwatch.cpp:35:  Do not leave a blank line after "public:"  [whitespace/blank_line] [3]
4: /github/home/ws/src/ros_gz/ros_ign_gazebo/src/Stopwatch.cpp:64:  Do not leave a blank line after "public:"  [whitespace/blank_line] [3]
4: /github/home/ws/src/ros_gz/ros_ign_gazebo/src/Stopwatch.cpp:67:  Do not leave a blank line after "public:"  [whitespace/blank_line] [3]
4: /github/home/ws/src/ros_gz/ros_ign_gazebo/src/Stopwatch.cpp:70:  Do not leave a blank line after "public:"  [whitespace/blank_line] [3]
4: /github/home/ws/src/ros_gz/ros_ign_gazebo/src/Stopwatch.cpp:73:  Do not leave a blank line after "public:"  [whitespace/blank_line] [3]
4: /github/home/ws/src/ros_gz/ros_ign_gazebo/src/Stopwatch.cpp:76:  Do not leave a blank line after "public:"  [whitespace/blank_line] [3]
```

```
The following tests FAILED:
	  4 - cpplint (Failed)
	  6 - uncrustify (Failed)
Errors while running CTest
--- stderr: ros_ign_gazebo
```

**Comment by andermi** [comment]
> Linters are failing

moving to gazebo-style with `public:` on each member caused the linters to fail

edit: I think I fixed it

**Review by ahcorde — APPROVED** [review]

**Pull request merged**